File: 2-add-two-numbers/add-two-numbers.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const dummy = new ListNode();
    let current = dummy;
    let carry = 0;

    while (l1 !== null || l2 !== null || carry !== 0) {
        let total = carry;

        if (l1 !== null) {
            total += l1.val;
            l1 = l1.next;
        }

        if (l2 !== null) {
            total += l2.val;
            l2 = l2.next;
        }

        const num = total % 10;
        carry = Math.floor(total / 10);

        current.next = new ListNode(num);
        current = current.next;
    }

    return dummy.next;
}

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode head = null;
        ListNode curHead = null;
        while(l1 != null || l2 != null || carry > 0) {
            int digit1 = (l1 != null) ? l1.val : 0;
            int digit2 = (l2 != null) ? l2.val : 0;
            int curSum = digit1 + digit2 + carry;
            carry = 0;
            l1 = (l1 == null) ? l1 : l1.next;
            l2 = (l2 == null) ? l2 : l2.next;
            int curDigit = curSum;
            if(curSum >= 10){
                carry = curSum / 10;
                curDigit = curSum % 10;
            }
            System.out.printf("curDigit : %d", curDigit);
            ListNode newNode = new ListNode(curDigit);
            if(curHead == null) {
                head = newNode;
                curHead = head;
            }
            else {
                curHead.next = newNode;
                curHead = curHead.next;
            }
        }
        return head;
    }
}